### DISCUSSION TITLE
Why the background task in RevalidatingServerAuthenticationStateProvider?

### DISCUSSION THREAD

**Discussion by DavidThielen** [dotnet/aspnetcore#49177] [Q&A]
Why wasn't RevalidatingServerAuthenticationStateProvider implemented as an override of GetAuthenticationStateAsync() where it has a LastValidated DateTime and compares that to DateTime.UtcNow to eighter return the stored authenticationStateTask or hit the DB to read in the latest?

I understand that this approach eliminates a call to DateTime.UtcNow on each call to GetAuthenticationStateAsync() but that's not an expensive call and GetAuthenticationStateAsync() is called just once for each new page being rendered.

Is there another reason for this approach? Or is it to improve every bit of performance everywhere (which is a good thing)?

thanks - dave

**Comment by gfoidl** [comment]
Does https://github.com/dotnet/aspnetcore/pull/12909 shed some light into why it's done this way?

**Reply by DavidThielen** [reply]
That is interesting - thank you. But I don't see anything in it that speaks to why it used a background task instead of overriding GetAuthenticationStateAsync(). Although it did allude to testability and so maybe that was part of why. If I missed something, please tell me.

I do appreciate that in it they stated they found UserManager was safe to call in Blazor code. I had found UserManager to work but was worried that it was not safe to use it. Glad to read it is.

thanks - dave

**Reply by gfoidl** [reply]
@SteveSandersonMS please have a look here and hopefully you can help clarify that question.
